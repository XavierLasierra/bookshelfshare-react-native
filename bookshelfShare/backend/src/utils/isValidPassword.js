const bcrypt = require("bcrypt");

function isValidPassword(password) {
  return bcrypt.compareSync(password, this.password);
}

module.exports = { isValidPassword };
