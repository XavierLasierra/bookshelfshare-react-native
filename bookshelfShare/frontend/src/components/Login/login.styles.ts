import {StyleSheet} from 'react-native';
import stylesConstants from '../../styles/styles.constants';

export default StyleSheet.create({
  topContainer: {
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center',
    height: 100,
  },
  loginContainer: {
    backgroundColor: stylesConstants.colors.light,
    borderTopRightRadius: 40,
    padding: stylesConstants.padding.normal,
    display: 'flex',
    flexDirection: 'column',
    alignItems: 'center',
    justifyContent: 'space-between',
    height: 600,
  },
  signUpTextContainer: {
    display: 'flex',
    flexDirection: 'column',
    alignItems: 'center',
  },
  loginButton: {
    marginTop: 20,
  },
  signUpText: {
    fontSize: stylesConstants.fontSize.big,
  },
  signUpButton: {
    fontSize: stylesConstants.fontSize.big,
    fontWeight: 'bold',
  },
});
