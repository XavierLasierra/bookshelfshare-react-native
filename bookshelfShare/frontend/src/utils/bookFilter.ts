export function bookFilter(text: string, books: any) {
  return books.reduce((acc: any, book: any) => {
    if (book.title.toLowerCase().includes(text.toLowerCase())) return [...acc, book];
    if (book.publisher.toLowerCase().includes(text.toLowerCase())) return [...acc, book];
    if (book.isbn.ISBN13.toLowerCase().includes(text.toLowerCase())) return [...acc, book];
    if (book.isbn.ISBN10.toLowerCase().includes(text.toLowerCase())) return [...acc, book];
    if (book.authors.some((author: string) => author.toLowerCase()
      .includes(text.toLowerCase()))) return [...acc, book];

    return acc;
  }, []);
}

export function bookShelfListFilter(text: string, books: any) {
  return books.reduce((acc: any, book: any) => {
    if (book.bookData.title.toLowerCase().includes(text.toLowerCase())) return [...acc, book];
    if (book.bookData.publisher.toLowerCase().includes(text.toLowerCase())) return [...acc, book];
    if (book.bookData.isbn.ISBN13.toLowerCase().includes(text.toLowerCase())) return [...acc, book];
    if (book.bookData.isbn.ISBN10.toLowerCase().includes(text.toLowerCase())) return [...acc, book];
    if (book.bookData.authors.some((author: string) => author.toLowerCase()
      .includes(text.toLowerCase()))) return [...acc, book];

    return acc;
  }, []);
}

export function bookShelfLocationFilter(filterLocation: number[], books: any) {
  return books.filter(({ customInformation: { location } }: any) => JSON.stringify(location)
  === JSON.stringify(filterLocation));
}
