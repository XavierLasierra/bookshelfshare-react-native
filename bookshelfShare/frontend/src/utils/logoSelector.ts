import BookIcon from '../assets/bookIcon.svg';
import SearchIcon from '../assets/searchIcon.svg';
import ProfileIcon from '../assets/profileIcon.svg';

export default function logoSelector(logoName: string) {
  switch (logoName) {
    case 'SearchIcon':
      return SearchIcon;
    case 'BookIcon':
      return BookIcon;
    case 'ProfileIcon':
      return ProfileIcon;
    default:
      return BookIcon;
  }
}
