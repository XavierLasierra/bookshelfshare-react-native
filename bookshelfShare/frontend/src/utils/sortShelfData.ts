export default function sortShelfData(shelf: any) {
  return shelf?.books.reduce((acc: any, book: any) => {
    const foundShelfBox = acc.find(({ location }: any) => JSON.stringify(location)
         === JSON.stringify(book?.customInformation?.location));
    if (foundShelfBox) {
      foundShelfBox.books.push(book.bookData);
    } else {
      acc.push({
        location: book?.customInformation?.location,
        books: [book.bookData]
      });
    }
    return acc;
  }, []);
}
