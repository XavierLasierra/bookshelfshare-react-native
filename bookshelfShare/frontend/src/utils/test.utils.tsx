import React from 'react';
import { render as rtlRender } from '@testing-library/react-native';
import { Provider } from 'react-redux';

import configureStore from '../redux/store';

function render(
  component: any,
  initialState: any
) {
  const store = configureStore(initialState);

  function Wrapper({ children }: any) {
    return (
      <Provider store={store}>
        {children}
      </Provider>
    );
  }
  return rtlRender(component, { wrapper: Wrapper });
}

// re-export everything
export * from '@testing-library/react-native';
// override render method
export { render };
