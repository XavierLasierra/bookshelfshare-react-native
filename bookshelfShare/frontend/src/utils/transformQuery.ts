interface Query {
    isbn?: string,
    inauthor?: string,
    intitle?: string,
    inpublisher?: string
}

export default function transformQuery(query: Query):string {
  const queryEntries = Object.entries(query);
  const url = queryEntries
    .reduce((acc, queryElement, index) => `${acc}${queryElement[0]}=${queryElement[1]}${index < queryEntries.length - 1 ? '+' : ''}`, '');
  return `${url}&key=${process.env.GOOGLE_API_KEY}`;
}
