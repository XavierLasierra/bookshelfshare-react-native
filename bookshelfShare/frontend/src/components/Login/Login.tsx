import React, { useState, useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import {
  View, Text, TextInput, TouchableOpacity, SafeAreaView, ActivityIndicator
} from 'react-native';
import { SharedElement } from 'react-navigation-shared-element';

import { validateEmail } from '../../utils/validation.utils';
import { loginUser } from '../../redux/actions/loggedUser.creator';

import BookIcon from '../../assets/bookIcon.svg';
import globalStyles from '../../styles/global.styles';
import styles from './login.styles';
import stylesConstants from '../../styles/styles.constants';

export default function Login({ navigation: { push } } : any) {
  const dispatch = useDispatch();
  const notification = useSelector((store: any) => store.notifications);
  const [userEmail, setUserEmail] = useState('');
  const [isValidEmail, setValidEmail] = useState(true);
  const [userPassword, setUserPassword] = useState('');
  const [isValidPassword, setValidPassword] = useState(true);
  const [isLoading, setIsLoading] = useState(false);

  useEffect(() => {
    setIsLoading(false);
  }, [notification]);

  function handleEmailChange(text: string) {
    setUserEmail(text);
  }

  function handleEmailFocus() {
    return !isValidEmail && setValidEmail(true);
  }

  function handlePasswordChange(text: string) {
    setUserPassword(text);
  }

  function handlePasswordFocus() {
    return !isValidPassword && setValidPassword(true);
  }

  function handleLogin() {
    if (validateEmail(userEmail)) {
      if (userPassword) {
        setIsLoading(true);
        dispatch(loginUser({
          email: userEmail,
          password: userPassword
        }));
      } else {
        setValidPassword(false);
      }
    } else {
      setValidEmail(false);
    }
  }

  function handleRegisterNavigation() {
    push('Register');
  }

  return (
    <SafeAreaView style={globalStyles.authenticationContainer}>
      <View style={styles.topContainer}>
        <SharedElement id="mainIcon">
          <BookIcon
            width={50}
            height={50}
          />
        </SharedElement>
      </View>
      <View style={styles.loginContainer}>
        <Text testID="loginTitle" style={globalStyles.title}>Log in</Text>
        <View style={globalStyles.inputContainer}>
          <Text style={globalStyles.inputLabel}>email</Text>
          <TextInput
            testID="emailInput"
            style={globalStyles.input}
            value={userEmail}
            onFocus={handleEmailFocus}
            onChangeText={handleEmailChange}
          />
          {!isValidEmail && <Text testID="invalidEmail" style={globalStyles.invalid}>Invalid email</Text>}
        </View>
        <View style={globalStyles.inputContainer}>
          <Text style={globalStyles.inputLabel}>password</Text>
          <TextInput
            testID="passwordInput"
            secureTextEntry
            style={globalStyles.input}
            value={userPassword}
            onFocus={handlePasswordFocus}
            onChangeText={handlePasswordChange}
          />
          {!isValidPassword && <Text testID="invalidPassword" style={globalStyles.invalid}>Type your password</Text>}
        </View>
        {!isLoading
          ? (
            <TouchableOpacity
              style={[globalStyles.button, styles.loginButton]}
              onPress={handleLogin}
              testID="loginButton"
            >
              <Text style={globalStyles.buttonText}>Log in</Text>
            </TouchableOpacity>
          )
          : (
            <TouchableOpacity
              style={[globalStyles.button, styles.loginButton]}
              testID="loginButton"
              disabled
            >
              <ActivityIndicator testID="loadingIndicator" size="large" color={stylesConstants.colors.white} />
            </TouchableOpacity>
          )}

        <View style={styles.signUpTextContainer}>
          <Text style={styles.signUpText}>Don&quot;t have an account?</Text>
          <Text
            style={styles.signUpButton}
            onPress={handleRegisterNavigation}
            testID="signUpButton"
          >
            Sign up
          </Text>
        </View>
      </View>
    </SafeAreaView>
  );
}
