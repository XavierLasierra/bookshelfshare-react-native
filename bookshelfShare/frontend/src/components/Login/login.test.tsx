import React from 'react';
import Login from './Login';
import { render, fireEvent } from '../../utils/test.utils';
import { loginUser } from '../../redux/actions/loggedUser.creator';
import notificationsActions from '../../redux/actions/notifications.actions';

const navigation = {
  push: jest.fn()
};

jest.mock('../../redux/actions/loggedUser.creator', () => ({
  loginUser: jest.fn()
}));

describe('Given a Login component', () => {
  describe('When it is rendered', () => {
    let screen: any;
    beforeEach(() => {
      screen = render(<Login navigation={navigation} />);
    });
    test('Then should match the snapshot', () => {
      const login = render(<Login navigation={navigation} />);
      expect(login).toMatchSnapshot();
    });

    describe('And email input text is changed', () => {
      test('Then the text typed should be rendered', () => {
        const emailInput = screen.getByTestId('emailInput');
        fireEvent(emailInput, 'focus');
        fireEvent.changeText(emailInput, 'email@email.com');
        expect(screen.queryByDisplayValue(/email@email.com/i)).not.toBe(null);
      });
    });

    describe('And password input text is changed', () => {
      test('Then the password typed should be rendered', () => {
        const passwordInput = screen.getByTestId('passwordInput');
        fireEvent(passwordInput, 'focus');
        fireEvent.changeText(passwordInput, 'mypassword');
        expect(screen.queryByDisplayValue(/mypassword/i)).not.toBe(null);
      });
    });

    describe('And login button is clicked', () => {
      describe('And the email is invalid', () => {
        beforeEach(() => {
          const loginButton = screen.getByTestId('loginButton');
          fireEvent.press(loginButton);
        });
        test('Then should render a Text with testID invalidEmail', () => {
          expect(screen.queryByTestId('invalidEmail')).not.toBe(null);
        });

        describe('And if after rendering invalidEmail you focus on email input', () => {
          test('Then should not render a Text with testId invalidEmail', () => {
            const emailInput = screen.getByTestId('emailInput');
            fireEvent(emailInput, 'focus');
            expect(screen.queryAllByTestId('invalidEmail').length).toBe(0);
          });
        });
      });

      describe('And the email is valid but the password is not', () => {
        beforeEach(() => {
          const emailInput = screen.getByTestId('emailInput');
          const loginButton = screen.getByTestId('loginButton');
          fireEvent.changeText(emailInput, 'email@email.com');
          fireEvent.press(loginButton);
        });

        test('Then should render a Text with testID invalidPassword', () => {
          expect(screen.queryByTestId('invalidPassword')).not.toBe(null);
        });

        describe('And if after rendering invalidPassword you focus on password input', () => {
          test('Then should not render a Text with testId invalidPassword', () => {
            const passwordInput = screen.getByTestId('passwordInput');
            fireEvent(passwordInput, 'focus');
            expect(screen.queryAllByTestId('invalidPassword').length).toBe(0);
          });
        });
      });

      describe('And the email and the password are valid', () => {
        describe('And the request isLoading', () => {
          beforeEach(() => {
            (loginUser as jest.Mock).mockReturnValue({
              type: ''
            });

            const emailInput = screen.getByTestId('emailInput');
            const passwordInput = screen.getByTestId('passwordInput');
            const loginButton = screen.getByTestId('loginButton');
            fireEvent.changeText(emailInput, 'email@email.com');
            fireEvent.changeText(passwordInput, 'myPassword');
            fireEvent.press(loginButton);
          });

          test('Then loginUser function should have been called', () => {
            expect((loginUser as jest.Mock)).toHaveBeenCalled();
          });

          test('Then loadingIndicator should be rendered', () => {
            expect(screen.queryByTestId('loadingIndicator')).not.toBe(null);
          });

          describe('And you click again while is loading', () => {
            test('Then loginUser should have not been called', () => {
              (loginUser as jest.Mock).mockClear();
              (loginUser as jest.Mock).mockReturnValue({
                type: ''
              });

              const loginButton = screen.getByTestId('loginButton');
              fireEvent.press(loginButton);
              expect((loginUser as jest.Mock).mock.calls.length).toBe(0);
            });
          });
        });

        describe('And the request is resolved (notification is sent)', () => {
          test('Then loadingIndicator should not be rendered', () => {
            (loginUser as jest.Mock).mockReturnValue({
              type: notificationsActions.LOGIN_ERROR
            });

            const emailInput = screen.getByTestId('emailInput');
            const passwordInput = screen.getByTestId('passwordInput');
            const loginButton = screen.getByTestId('loginButton');
            fireEvent.changeText(emailInput, 'email@email.com');
            fireEvent.changeText(passwordInput, 'myPassword');
            fireEvent.press(loginButton);

            expect(screen.queryByTestId('loadingIndicator')).toBe(null);
          });
        });
      });
    });

    describe('And sign up button is clicked', () => {
      test('Then navigation.push should have been called with Register', () => {
        const signUpButton = screen.getByTestId('signUpButton');
        fireEvent.press(signUpButton);
        expect(navigation.push).toHaveBeenCalledWith('Register');
      });
    });
  });
});
