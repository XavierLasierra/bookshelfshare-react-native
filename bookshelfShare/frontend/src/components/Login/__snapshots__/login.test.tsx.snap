// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given a Login component When it is rendered Then should match the snapshot 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "backgroundColor": "#D9D8C4",
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "height": 100,
        "justifyContent": "center",
      }
    }
  >
    <View
      collapsable={false}
    >
      <IconMock
        height={50}
        width={50}
      />
    </View>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#F5F5F5",
        "borderTopRightRadius": 40,
        "display": "flex",
        "flexDirection": "column",
        "height": 600,
        "justifyContent": "space-between",
        "padding": 30,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#3E3C37",
          "fontSize": 35,
          "fontWeight": "500",
        }
      }
      testID="loginTitle"
    >
      Log in
    </Text>
    <View
      style={
        Object {
          "height": 70,
          "width": "100%",
        }
      }
    >
      <Text
        style={
          Object {
            "fontSize": 12,
            "fontWeight": "bold",
            "transform": Array [
              Object {
                "translateX": 5,
              },
              Object {
                "translateY": 20,
              },
            ],
            "zIndex": 1,
          }
        }
      >
        email
      </Text>
      <TextInput
        onChangeText={[Function]}
        onFocus={[Function]}
        style={
          Object {
            "backgroundColor": "#FFFFFF",
            "borderColor": "#ECECEC",
            "borderRadius": 10,
            "borderWidth": 1,
            "fontSize": 20,
            "height": "100%",
            "paddingHorizontal": 30,
            "paddingTop": 30,
          }
        }
        testID="emailInput"
        value=""
      />
    </View>
    <View
      style={
        Object {
          "height": 70,
          "width": "100%",
        }
      }
    >
      <Text
        style={
          Object {
            "fontSize": 12,
            "fontWeight": "bold",
            "transform": Array [
              Object {
                "translateX": 5,
              },
              Object {
                "translateY": 20,
              },
            ],
            "zIndex": 1,
          }
        }
      >
        password
      </Text>
      <TextInput
        onChangeText={[Function]}
        onFocus={[Function]}
        secureTextEntry={true}
        style={
          Object {
            "backgroundColor": "#FFFFFF",
            "borderColor": "#ECECEC",
            "borderRadius": 10,
            "borderWidth": 1,
            "fontSize": 20,
            "height": "100%",
            "paddingHorizontal": 30,
            "paddingTop": 30,
          }
        }
        testID="passwordInput"
        value=""
      />
    </View>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      nativeID="animatedComponent"
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#000000",
          "borderRadius": 10,
          "display": "flex",
          "height": 70,
          "justifyContent": "center",
          "marginTop": 20,
          "opacity": 1,
          "width": "100%",
        }
      }
      testID="loginButton"
    >
      <Text
        style={
          Object {
            "color": "#FFFFFF",
            "fontSize": 20,
            "fontWeight": "500",
          }
        }
      >
        Log in
      </Text>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "display": "flex",
          "flexDirection": "column",
        }
      }
    >
      <Text
        style={
          Object {
            "fontSize": 20,
          }
        }
      >
        Don"t have an account?
      </Text>
      <Text
        onPress={[Function]}
        style={
          Object {
            "fontSize": 20,
            "fontWeight": "bold",
          }
        }
        testID="signUpButton"
      >
        Sign up
      </Text>
    </View>
  </View>
</RCTSafeAreaView>
`;
